SCAutolib/controller.py

```diff
@@ -0,0 +1,375 @@
+        schema_cas = Schema(And(
+            Use(dict),
+            # Check that CA section contains at least one and maximum
+            # two entries
+            lambda l: 1 <= len(l.keys()) <= 2,
+            {Optional("local_ca"): {"dir": Use(Path)},
+             Optional("ipa"): {
+                 'admin_passwd': Use(str),
+                 'root_passwd': Use(str),
+                 Optional('ip_addr', default=None): Use(str),
+                 'server_hostname': Use(str),
+                 'client_hostname': Use(str),
+                 'domain': Use(str),
+                 'realm': Use(str.upper)}}),
+            ignore_extra_keys=True)
+
+        # Specify validation schema for all users
+        schema_user = Schema({'name': Use(str),
+                              'passwd': Use(str),
+                              'pin': Use(str),
+                              Optional('card_dir', default=None): Use(Path),
+                              'card_type': Or("virtual", "real", "removinator"),
+                              Optional('cert', default=None): Use(Path),
+                              Optional('key', default=None): Use(Path),
+                              'local': Use(bool)})
+
+        # Specify general schema for whole config file
+        schema = Schema({"root_passwd": Use(str),
+                         "ca": schema_cas,
+                         "users": [schema_user]})
+
+        return schema.validate(self.lib_conf)
+
+    @staticmethod
+    def _general_steps_for_virtual_sc():
+        """
+        Prepare the system for virtual smart card. Preparation means to
+        configure pcscd service and opensc module to be able correctly working
+        with virtual smart card. Also, repository for installing virt_cacard
+        package is added in this method.
+        """
+
+        _check_selinux()
+
+        pcscd_service = File("/usr/lib/systemd/system/pcscd.service")
+        pcscd_service.backup()
+        exec_start = pcscd_service.get(section="Service", key="ExecStart")
+        if "--auto-exit" in exec_start:
+            exec_start = exec_start.replace("--auto-exit", "")
+            pcscd_service.set(section="Service", key="ExecStart",
+                              value=exec_start)
+            pcscd_service.save()
+
+        opensc_module = File("/usr/share/p11-kit/modules/opensc.module")
+        opensc_module.backup()
+        try:
+            opensc_module.get("disable-in", separator=":")
+        except SCAutolibException:
+            logger.warning("OpenSC module do not have option 'disabled-in: "
+                           "virt_cacard' set")
+            opensc_module.set(key="disabled-in", value="virt_cacard",
+                              separator=": ")
+            opensc_module.save()
+
+        run(['systemctl', 'stop', 'pcscd.service', 'pcscd.socket', 'sssd'])
+        rmtree("/var/lib/sss/mc/*", ignore_errors=True)
+        rmtree("/var/lib/sss/db/*", ignore_errors=True)
```

Comment:
Are these directories something that we can safely remove?

Reply:
yes. it is cache directories